drivers/platform/aducm3029/flash.c

```diff
@@ -0,0 +1,250 @@
+/******************************************************************************/
+
+#include <drivers/flash/adi_flash.h>
+#include <drivers/dma/adi_dma.h>
+#include <stdlib.h>
+#include <malloc.h>
+#include "flash.h"
+#include "flash_extra.h"
+#include "error.h"
+
+/******************************************************************************/
+/*************************** Types Declarations *******************************/
+/******************************************************************************/
+
+/**
+ * @struct adicup_flash_dev
+ * @brief Aducm3029 flash controller handler
+ */
+struct adicup_flash_dev {
+	void *instance;
+	void *dfp_memory;
+};
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+
+/**
+ * Initialize flash controller.
+ *
+ * @param [out] device     - Pointer to the driver handler.
+ * @param [in]  init_param - Pointer to the initialization structure.
+ *
+ * @return 0 in case of success, error code otherwise.
+ */
+int32_t flash_init(struct flash_dev **device,
+		   struct flash_init_param *init_param)
+{
+	int32_t ret;
+	struct flash_dev *dev;
+	struct adicup_flash_dev *adicup_extra;
+
+	dev = (struct flash_dev *)calloc(1, sizeof(*dev));
+	if (!dev)
+		return FAILURE;
+
+	dev->id = init_param->id;
+	adicup_extra = (struct adicup_flash_dev *)calloc(1,
+			sizeof(*adicup_extra));
+	if (!adicup_extra)
+		goto error_dev;
+
+	dev->extra = adicup_extra;
+	adicup_extra->dfp_memory = calloc(1, ADI_FEE_MEMORY_SIZE);
+	if (!adicup_extra->dfp_memory)
+		goto error_adicup;
+
+	/* Open flash device */
+	ret = adi_fee_Open(dev->id, adicup_extra->dfp_memory,
+			   ADI_FEE_MEMORY_SIZE,
+			   (ADI_FEE_HANDLE *)&adicup_extra->instance);
+	if(ret != SUCCESS)
+		goto error_dfp_mem;
+
+	*device = dev;
+
+	return SUCCESS;
+
+error_dfp_mem:
+	free(adicup_extra->dfp_memory);
+error_adicup:
+	free(adicup_extra);
+error_dev:
+	free(dev);
+
+	return FAILURE;
+}
+
+/**
+ * Free memory allocated by flash_init().
+ *
+ * @param [in] dev - Pointer to the driver handler.
+ *
+ * @return 0 in case of success, error code otherwise.
+ */
+int32_t flash_remove(struct flash_dev *dev)
+{
+	int32_t ret;
+	struct adicup_flash_dev *adicup_extra;
+
+	if (!dev)
+		return FAILURE;
+
+	adicup_extra = dev->extra;
+
+	ret = adi_fee_Close(adicup_extra->instance);
+	if (ret != SUCCESS)
+		return FAILURE;
+
+	free(adicup_extra->dfp_memory);
+	free(adicup_extra);
+	free(dev);
+
+	return SUCCESS;
+}
+
+/**
+ * Read-modify-write helper function for flash_write(). Since the smallest write
+ * unit for the ADuCM3029 is the page this function helps access data smaller than
+ * a page or accross multiple pages.
+ */
+static int32_t flash_write_page(struct flash_dev *dev, uint32_t flash_addr,
+				uint32_t *array, uint32_t array_size)
```

Comment:
If writting a page why array_size is needed?

Reply:
Maybe should rename. This takes care of a write within a page, but we still need to know how much we are writing. What this does is:

1. Read one page (in temporary buffer);
2. Change only the values we want inside the buffer (this is where we use the size);
3. Erase page;
4. Write temporary (size of one page) buffer to page.

Reply:
I would do something like: `flash_write_in_page(struct flash_dev *dev, uint32_t page, uin32_t offset, uint32_t *data, uint32_t len)`

Reply:
Renamed the function to something more intuitive.